### PR TITLE
ingest: support default pipelines + bulk upserts

### LINKED ISSUE #36219: ingest: Bulk update does not honor default ingest node pipeline
6.5. introduced the ability to define a default ingest node pipeline for a given index. 

However, the default pipeline is not honored when implementing a bulk update index request. It is however honored for non-bulk updates.

For example:
```
DELETE test_index
PUT _ingest/pipeline/test_pipeline
{
    "processors" : [
      {
        "set": {
          "field": "ran_pipeline",
          "value": true
        }
      }
    ]
}

PUT test_index
{
  "settings": {
    "default_pipeline": "test_pipeline"
  }
}
POST _bulk
{"update":{"_id":"1","_index":"test_index","_type":"_doc","_routing":null,"retry_on_conflict":1}}
{"script":{"source":"ctx._source.ran_script = true"},"upsert":{"ran_upsert":true}}

POST _bulk
{"update":{"_id":"1","_index":"test_index","_type":"_doc","_routing":null,"retry_on_conflict":1}}
{"script":{"source":"ctx._source.ran_script = true"},"upsert":{"ran_upsert":true}}

POST _bulk
{"update":{"_id":"1","_index":"test_index","_type":"_doc","_routing":null,"retry_on_conflict":1}}
{ "doc" : {"ran_doc" : true}, "_source": true}

GET test_index/_doc/1
```
Results in 
```
{
  "_index" : "test_index",
  "_type" : "_doc",
  "_id" : "1",
  "_version" : 3,
  "found" : true,
  "_source" : {
    "ran_upsert" : true,
    "ran_script" : true,
    "ran_doc" : true
  }
}
```
^^ Note the missing "ran_pipeline" 

Compared to the non-bulk API call which works as expected:
```
POST test_index/_doc/2/_update
{
  "script":{
    "source": "ctx._source.ran_script = true" 
  },
  "upsert" :{
    "ran_upsert" : true
  }
}
POST test_index/_doc/2/_update
{
  "script":{
    "source": "ctx._source.ran_script = true" 
  },
  "upsert" :{
    "ran_upsert" : true
  }
}
POST test_index/_doc/2/_update
{
  "doc":{
    "ran_doc" : true
  }
}
GET test_index/_doc/2
```
Results in 
```
{
  "_index" : "test_index",
  "_type" : "_doc",
  "_id" : "2",
  "_version" : 3,
  "found" : true,
  "_source" : {
    "ran_script" : true,
    "ran_upsert" : true,
    "ran_pipeline" : true,
    "ran_doc" : true
  }
}
```
^^ Correctly ran the pipeline.

### PR DESCRIPTION
**Pull request by jakelandis**
This commit adds support to enable bulk upserts to use an index's
default pipeline. Bulk upsert, doc_as_upsert, and script_as_upsert
are all supported.

However, bulk script_as_upsert has slightly surprising behavior since
the pipeline is executed _before_ the script is evaluated. This means
that the pipeline only has access the data found in the upsert field
of the script_as_upsert. The non-bulk script_as_upsert (existing behavior)
runs the pipeline _after_ the script is executed. This commit
does _not_ attempt to consolidate the bulk and non-bulk behavior for
script_as_upsert.

This commit also adds additional testing for the non-bulk behavior,
which remains unchanged with this commit.

fixes #36219

### CODE REVIEW

**Comment by original-brownbear** [comment]
@jakelandis can you take a look at test failures here?

This one looks new and possible related to your work: 

```
23:41:34 ERROR   0.31s J6 | BulkWithUpdatesIT.testBulkUpdateSimple <<< FAILURES!
23:41:34    > Throwable #1: java.lang.NullPointerException
23:41:34    > 	at __randomizedtesting.SeedInfo.seed([A74EC1584801E727:6F580076C91AFCB8]:0)
23:41:34    > 	at java.util.TreeMap.getEntry(TreeMap.java:347)
23:41:34    > 	at java.util.TreeMap.get(TreeMap.java:278)
23:41:34    > 	at java.util.Collections$UnmodifiableMap.get(Collections.java:1454)
23:41:34    > 	at org.elasticsearch.action.bulk.TransportBulkAction.executeIngestAndBulk(TransportBulkAction.java:235)
23:41:34    > 	at org.elasticsearch.action.bulk.TransportBulkAction.doExecute(TransportBulkAction.java:183)
23:41:34    > 	at org.elasticsearch.action.bulk.TransportBulkAction.doExecute(TransportBulkAction.java:89)
23:41:34    > 	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:141)
23:41:34    > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:117)
23:41:34    > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:60)
23:41:34    > 	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:83)
23:41:34    > 	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:72)
23:41:34    > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:393)
23:41:34    > 	at org.elasticsearch.client.FilterClient.doExecute(FilterClient.java:65)
23:41:34    > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:393)
23:41:34    > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:382)
23:41:34    > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:45)
23:41:34    > 	at org.elasticsearch.action.ActionRequestBuilder.get(ActionRequestBuilder.java:52)
23:41:34    > 	at org.elasticsearch.action.bulk.BulkWithUpdatesIT.testBulkUpdateSimple(BulkWithUpdatesIT.java:168)
23:41:34    > 	at java.lang.Thread.run(Thread.java:748)
```

(from https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/2302/console)

**Comment by jakelandis** [comment]
@elasticmachine run gradle build 2

**Comment by jakelandis** [comment]
@elasticmachine run gradle build tests 2

**Comment by jakelandis** [comment]
@original-brownbear  test failure fixed. Ready for review.

File: modules/ingest-common/src/test/resources/rest-api-spec/test/ingest/200_default_pipeline.yml [hunk 1]
```diff
@@ -63,28 +63,117 @@ teardown:
...
+  - match: { _source.bytes_target_field: 1024 }
+# default pipeline via bulk upsert
+# note - bulk scripted upsert's execute the pipeline before the script, so any data referenced by the pipeline
+# needs to be in the upsert, not the script
+  - do:
+      bulk:
+        refresh: true
+        body: |
+          {"update":{"_id":"6","_index":"test","_type":"test"}}
+          {"script":{"source":"ctx._source.ran_script = true"},"upsert":{"bytes_source_field":"1kb"}}
```
Comment (original-brownbear) [inline modules/ingest-common/src/test/resources/rest-api-spec/test/ingest/200_default_pipeline.yml]: NIT: You should be able to just shorten this (`"script":{"source":"ctx._source.ran_script = true"}`) to `"script":"ctx._source.ran_script = true"`, no need for the object when using the defaults+painless :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 2]
```diff
@@ -566,6 +584,7 @@ void processBulkIndexIngestRequest(Task task, BulkRequest original, ActionListen
             indexRequest -> bulkRequestModifier.markCurrentItemAsDropped());
     }
 
+
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: NIT: Noisy new line

File: server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionTests.java [hunk 3]
```diff
@@ -132,4 +134,24 @@ public void testDeleteNonExistingDocExternalGteVersionCreatesIndex() throws Exce
...
+        assertEquals(TransportBulkAction.getIndexWriteRequest(upsertRequest), indexRequest);
+        assertEquals(TransportBulkAction.getIndexWriteRequest(docAsUpsertRequest), indexRequest);
+        assertEquals(TransportBulkAction.getIndexWriteRequest(scriptedUpsert), indexRequest);
+
+        DeleteRequest deleteRequest = new DeleteRequest("index", "id");
+        assertNull(TransportBulkAction.getIndexWriteRequest(deleteRequest));
+
+        UpdateRequest badUpsertRequest = new UpdateRequest("index", "type", "id1");
+        assertNull(TransportBulkAction.getIndexWriteRequest(badUpsertRequest));
+
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionTests.java]: NIT: Random empty line :)

**Review by original-brownbear — APPROVED** [review]
LGTM :)

BTW, 2 comments :
1. I think we should really consolidate the execution order between pipeline and script upsets in the bulk vs. non-bulk case.
Also, I wonder what happens (or should happen) if the pipeline ran into the `drop` processor for a scripted upsert?

**Comment by jakelandis** [comment]
@elasticmachine run gradle build tests 2

**Comment by jakelandis** [comment]
@original-brownbear - thanks for the review. I have logged https://github.com/elastic/elasticsearch/issues/36745 to adress the execution order between bulk and non-bulk, and https://github.com/elastic/elasticsearch/issues/36746 to address a bug with upsert and drop.

**Pull request merged**